### ISSUE TITLE
problem in opening .mat model with cobraPy

### ISSUE THREAD

**Issue by Ilhem36** [opencobra/cobrapy#1190]
#### Problem description

Hello every one i'm trying to open a .mat model with cobraPy it was running for a week in Jupyter but I received an error after one week.can you please explain to me how can I  fix this error ? 
Thank You very much  I

Please explain:
* **what** I'm trying to open .mat model with cobraPy,
* **how** you went about it (referring to the code sample), and
* **why** the current behaviour is a problem and what output
  you expected instead.

#### Code Sample

Create a [minimal, complete, verifiable example](https://stackoverflow.com/help/mcve).

```python
```
UserWarning: malformed gene_reaction_rule '[]' for <Reaction Blautia_hydrogenotrophica_DSM_10507_IEX_val_L[u]tr at 0x7f3dfae24860>
  warn("malformed gene_reaction_rule '%s' for %s" % (new_rule, repr(self)))
/opt/python-3.6.9/lib/python3.6/site-packages/cobra/core/reaction.py:438: UserWarning: malformed gene_reaction_rule '[]' for <Reaction Blautia_hydrogenotrophica_DSM_10507_IEX_zn2[u]tr at 0x7f3dfae248d0>
  warn("malformed gene_reaction_rule '%s' for %s" % (new_rule, repr(self)))
/opt/python-3.6.9/lib/python3.6/site-packages/cobra/core/reaction.py:438: UserWarning: malformed gene_reaction_rule '[]' for <Reaction Blautia_hydrogenotrophica_DSM_10507_IEX_no3[u]tr at 0x7f3dfae24940>
  warn("malformed gene_reaction_rule '%s' for %s" % (new_rule, repr(self)))
IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)

```
```

### Context

<!-- Please run the following code and paste the output between the ``` tickmarks below
inside the details block.

python -c "import cobra;cobra.show_versions()"

model_micro = cobra.io.load_matlab_model('microbiota_model_samp_SRS011239.mat')

**Comment by cdiener** [comment]
Is that multi-taxon model? Because it sounds like it might be a huge model that crashes jupyter. As the error states there are also invalid GPRs in the model. If that is taken from the microbiome cobratoolbox this probably has to be reported there.

**Comment by Ilhem36** [comment]
Hello , Thank for your answer. 
Yes the model was generated by cobratoolbox, but I don't have enough information about it . I think that is a microbiome model for 149 healthy individuals according to the article. it was opened rapidly by octave but by cobraPy it was turning for a  week and it crashed. Do you think that I can correct invalid GRPS  in the model using cobrapy because i can't use cobratoolbox .
So I changed the python version and I have this error now : 

```
Scaling...
 A: min|aij| =  1.000e+00  max|aij| =  1.000e+00  ratio =  1.000e+00
Problem data seem to be well scaled
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_49380/47059282.py in <module>
----> 1 microbiome=cobra.io.load_matlab_model("microbiota_model_samp_SRS011239.mat")

~/.local/lib/python3.7/site-packages/cobra/io/mat.py in load_matlab_model(infile_path, variable_name, inf)
     81             variable_name = possible_names[0]
     82     if variable_name is not None:
---> 83         return from_mat_struct(data[variable_name], model_id=variable_name, inf=inf)
     84     for possible_name in possible_names:
     85         try:

~/.local/lib/python3.7/site-packages/cobra/io/mat.py in from_mat_struct(mat_struct, model_id, inf)
    239             coefficients[new_reaction] = float(c_vec[i][0])
    240         try:
--> 241             new_reaction.gene_reaction_rule = str(m["grRules"][0, 0][i][0][0])
    242         except (IndexError, ValueError):
    243             pass

~/.local/lib/python3.7/site-packages/cobra/core/reaction.py in gene_reaction_rule(self, new_rule)
    489             warn("Context management not implemented for gene reaction rules.")
    490 
--> 491         self._gpr = GPR.from_string(new_rule)
    492         self._update_genes_from_gpr()
    493 

~/.local/lib/python3.7/site-packages/cobra/core/gene.py in from_string(cls, string_gpr)
    360                 warn(e.msg)
    361                 return gpr
--> 362         return cls(tree)
    363 
    364     @property

~/.local/lib/python3.7/site-packages/cobra/core/gene.py in __init__(self, gpr_from, **kwargs)
    300                 # noinspection PyTypeChecker
    301                 self.body = deepcopy(gpr_from.body)
--> 302                 self.eval()
    303             else:
    304                 raise TypeError("GPR requires AST Expression or Module")

~/.local/lib/python3.7/site-packages/cobra/core/gene.py in eval(self, knockouts)
    453             knockouts = list(knockouts)
    454         if self.body:
--> 455             return self._eval_gpr(self.body, knockouts=knockouts)
    456         else:
    457             return True

~/.local/lib/python3.7/site-packages/cobra/core/gene.py in _eval_gpr(self, expr, knockouts)
    428             return True
    429         else:
--> 430             raise TypeError(f"Unsupported operation: {repr(expr)}")
    431 
    432     def eval(self, knockouts: Union[DictList, Set, str, Iterable] = None) -> bool:

TypeError: Unsupported operation: <_ast.List object at 0x7f61df434950>
```

But I can't understand how to fix it.
Thank you.

**Comment by cdiener** [comment]
It's the same error (the error message is a bit obtuse here). It complains because you have gene reaction rules that are "[]". If you wanted to fix it you could just load the mat file with sicpy and than fix the wrong GPRs by hand, write it to a new mat file and read it. But in the end the problem seems to be cobratoolbox writing invalid GPRs and we can't really fix this from our side.

**Comment by akaviaLab** [comment]
I think it will be fixed when #1197 is merged into the tree. This was something I rewrote, to have empty string instead of empty list for gene rules.

**Comment by Ilhem36** [comment]
Can you show me please how did you fix it . Because I tried with scipy to write something in the list instead of having empty list, and I saveed the model but it didn't work. Thank you very much for your help.

**Comment by akaviaLab** [comment]
I wrote several patches in the PR request. You can follow the link. If you don't understand how to do that, just wait.

You can also give me a link to the WBM model, and I'll try to make it work. I don't know which model you are using, and I would need a link to check your model.

**Comment by Ilhem36** [comment]
https://www.thielelab.eu/in-silico-models   Here is the link where I downloaded the WBM model but  you have to fix an error manually to open  the model with cobrapy.

**Comment by akaviaLab** [comment]
Which one of the models in the website is the relevant one?
Please post a link to the mat file that is causing *this* error.

**Comment by Ilhem36** [comment]
https://www.thielelab.eu/in-silico-models here is the link ( Harvey model ). 
https://www.thielelab.eu/in-silico-models , you have just to click on harvey in this link and the model will be downloaded. 
The mat file which is causing this error is the  same as in this link but I modified it to be opened by cobrapy because there is an error and I can't find where can I add the correct .mat file in this comment.

**Comment by cdiener** [comment]
You can zip the model and drag it into the edit box or click on `Attach files by dragging & dropping, selecting or pasting them` at the bottom of the edit box.

**Comment by akaviaLab** [comment]
Some of the models, assuming you're trying to load the joint, like modelJoint_Enterocyte_LPSTy.mat have a weird situation where the dimensions of some fields (in this case 'metCharge') is different from the number of mets.
3832 mets
773 metCharge

This might be the error, but it is dependent on which model is problematic (which you still haven't told us).

If this is the case, I don't think we'll fix it. Please contact the people who made the original mat file.

**Comment by cdiener** [comment]
In your case it looks like you're missing compartments actually. We might want to check for this case in the external compartment heuristic if we don't already. If you downloaded the models from the links mentioned in https://github.com/opencobra/cobrapy/issues/1194 those will be fixed by @akaviaLab 's PR as you can see there (with the exchange detection working).

**Comment by Ilhem36** [comment]
[Harvey4.zip](https://github.com/opencobra/cobrapy/files/8444380/Harvey4.zip)
Hello here is the model where I'm working on.

**Comment by akaviaLab** [comment]
Loads without problem on my branch. Should work for you once #1197 is merged.

**Comment by Ilhem36** [comment]
Yes I can also load the model but I asked about how to acess some  information ( such us external compartments, exchange , sink and demand reactions) using cobrapy functions in this model.
 For instance I used : find_external_compartment(model_harvey) but I got errors (you find it above) but when I use it with .json model , it is working. Do you think that there is another manner to access information in  this model? .
In fact my goal is to merge two models the WBM(.mat file) and the microbiote community (.json file), but I don't know if it is possible with cobrapy ? or not ?

**Comment by cdiener** [comment]
The exchanges etc are working on that brunch, see my comment above. Merging the model is possible but there is no simple way and will need some manual coding to do so. If both models share a compartment and use the same ID space it may be as simple as doing `model1.add_reactions(model2.reactions)` or something more complicated. The size of the model can also be an issue. The solver settings defaults don't work well for models with more than 50k reactions. Tools like micom are a bit more specialized there but it does not support host-microbiome models yet (though we are working on it).

**Comment by akaviaLab** [comment]
Okay.
The harvey4 model is weird, since it has some compartments that have
capital letters.
I can make it work if I change the regular expression in cobra.io.mat to
include capitals.

However, I'm a bit hesitant to do so, since it might pick up weird
metabolite names, and harvey4 seems unique. @cdiener - thoughts?

As for the other question on merging models - no idea, it is probably
possible, but would be hard.

**Comment by akaviaLab** [comment]
Exchanges might work, but find_external_compartments will fail on the
harvey4, as provided.
It will fail since some mets have compartments with capital letters, which
is not picked up by the regex.
These compartments with capital letters are marked as None, which seems
problematic for metabolites.
Hence my question.

**Comment by cdiener** [comment]
Hmm, I will have a look. Does the Matlab format not have compartment assignments for metabolites? Getting the via a regex can be problematic since that is not a very strict convention.

**Comment by akaviaLab** [comment]
I went over it, and I can fix it.
However, there are some metabolites that look like this xylan[luSI], which
is not picked up by the regular expression. Matlab doesn't have to have
explicit compartments. It can just be part of met id (unfortunately). The
matlab format CAN have explicit compartments, but usually doesn't.

If you extend the regular expression to deal with capitals, it works. I'm
just worried if it might have weird side effects on other models, which are
more common than Harvey. Maybe not, because the regex includes $ (end of
metabolite id).

**Comment by cdiener** [comment]
Yeah that makes sense. Maybe also make sure that it's only a single underscore. Some community models use things like "met_e__E_coli" which looks like it would trip up the regex

**Comment by akaviaLab** [comment]
That wouldn't work because the underscore (and the bracket) have $ which is
the end of the metabolite

"met_e__E_coli" might pick up coli with the regex re.compile(r"
_(?P<compartment>[a-zA-Z]+)$")
I can't think of a regex that will pick up the compartment from that, and
not get mixed up with something like lansterol_13_c

**Comment by cdiener** [comment]
I'm really not sure whether we should read Matlab models that define no compartments whatsoever. Is that the case here or do they at least have a list of all compartment which we could sub into the regex? I would argue that a model with zero compartment information does not have sufficient information to be supported by cobrapy at all. Or we could just assign a generic compartment to all metabolites in that case. This would break the exchange detection but you can't define exchanges without compartments so that is out of our responsibility.

**Comment by akaviaLab** [comment]
Sorry.
Harvey 4 has compartments.
The compartments are a bit weird, since they have capital letters. I added
something to the regex to do pick up capital letters as well. I'm a bit
worried this new regex might get confused by metabolite names that have
capitals, but it doesn't seem too likely.

I can't think of a regex that will pick up your example "met_e__E_coli"
and not get confused by metabolite names with underscores (of which there
are many).

I also added a check that will raise an exception if there are empty or
None compartments when reading from matlab. I agree we shouldn't read that
kind of model and it should be fixed externally.

**Comment by cdiener** [comment]
So if there is a list of compartments we could just put them in the regex explicitly so it only detects those names in IDs. If it's only via names there will always be issues, for instance D-Glucose is often named "glc_D" or "glc__D" and this would also create false positives. It's definitely much easier if you know the compartment names and can look specifically for "_e" or "_c" in the IDs.

**Comment by akaviaLab** [comment]
The compartments in Harvey4 are in the format of []
Yes, glc_D might give false positives, but only if it doesn't have
compartments.
So glc__D_e would pick "e" as the compartment.
If some metabolite is just glc_D, not glc_D_e or glc_D[e] then it doesn't
have compartments defined, so that seems less of a problem.

**Comment by cdiener** [comment]
Ufff, disappointing that they provide no compartments at all. We should definitely warn if that is the case.

> Yes, glc_D might give false positives, but only if it doesn't have
compartments.

I mean if the model has no compartments and we add in a bunch of random ones that does not sound great to me.

**Comment by akaviaLab** [comment]
Okay.
I feel like we are not fully communicating.

Harvey4 has compartments as part of the met id, but not explicitly.  The
original cobra.io.mat functions dealt with this, using regex. I just
rewrote them to allow capital letters in compartments, and reorganized some
of the other functions. There are two regex - one dealing with the
format _COMPARMENT and one dealing with the format [COMPARTMENT]

If the mat model has compartments explicitly, as a different MATLAB field,
those will be used, and the met id will not be processed with regex. If
there are no explicit compartments, the met id will be used to infer them.

Your example, where the compartment id is in the middle of the met id will
lead to confusion. "met_e__E_coli" will trip up the regex, and lead to the
compartment being identified as "coli". If this kind of model has an
explicit compartment, it won't be a problem.

If a model without compartments has "glc_D", then *that* metabolite will be
identified as belonging to compartment D. However, there probably will be
other metabolites, such as "atp" which will have no compartments. A model
that has empty string or None for ANY metabolite compartments will fail
with a ValueError.
If a model without compartments has "glc_D_e" or "glc_D[e]", the
compartment will be "e".
Likewise if the model without compartment has
"glc_D_longNameForComparment", the short compartment id will be identified
as "longNameForCompartment". That's possibly silly, but not random.

I hope this covered the issues.

**Comment by cdiener** [comment]
Okay, if it fails on any metabolite with missing compartment that sounds fine. So that means the Harvey model failed even in your PR because the compartment with the upper case compartment was not assigned before you adjusted the regex, right? So no model ever returned by `load_matlab_model` has metabolites with missing compartments?

**Comment by akaviaLab** [comment]
So that means the Harvey model failed even in your PR because the
compartment with the upper case compartment was not assigned before you
adjusted the regex, right?  That was the case with harvey4.

So no model ever returned by load_matlab_model has metabolites with missing
compartments?
Well, it looks like harvey4 is somewhat of an exception.
It has some fake metabolites that have no defined compartment, and are of
the format
'Scord_biomass_maintenance_dummy_objective'

and are accumulation of biomass components in Scord (spinal cord?) tissue.

The mat loading function uses two regex, and most metabolites in harvey4
are of the format met[Compartment]. That kind of compartment is now picked
up correctly after I adjusted the regex.
The biomass metabolites of the format
'Scord_biomass_maintenance_dummy_objective' don't match met[Comaprtment],
but they do match the met_Compartment regex. Then they are assigned to the
fake compartment 'objective'. The functions check if either regex matches.

I could modify the functions so that all metabolites must have one format
or another. But that seems like the code would be a bit complicated, and
would mostly result in failing harvey4. It would also result in a much
stricter mat import, and I'm not sure that cobratoolbox enforces this
strict format adherence.
I don't think any other models have this weird mixture of metabolites from
two regex, without explicit compartments.

**Comment by cdiener** [comment]
Hmm, I think it's common metabolite IDs have underscores, so we will often get false positives. The alternative is to assign a common compartment and give a warning that the user should revise the compartment assignments. 

Though I think your solution is fine, maybe just trigger a logger warning that the compartments have deduced heuristically and that the user should check the compartments manually afterwards.

**Comment by akaviaLab** [comment]
Only if the underscores are at the end of the id and there are no defined
compartments. Not that common

**Comment by cdiener** [comment]
Yep agreed.

**Comment by Ilhem36** [comment]
> I wrote several patches in the PR request. You can follow the link. If you don't understand how to do that, just wait.
> 
> You can also give me a link to the WBM model, and I'll try to make it work. I don't know which model you are using, and I would need a link to check your model.



> The exchanges etc are working on that brunch, see my comment above. Merging the model is possible but there is no simple way and will need some manual coding to do so. If both models share a compartment and use the same ID space it may be as simple as doing `model1.add_reactions(model2.reactions)` or something more complicated. The size of the model can also be an issue. The solver settings defaults don't work well for models with more than 50k reactions. Tools like micom are a bit more specialized there but it does not support host-microbiome models yet (though we are working on it).

for me exchanges doesn't work on havey4 because when I wrote model.exchanges or model.sinks or model.demands it gave me this error:  "IndexError: tuple index out of range".

**Comment by Ilhem36** [comment]
> I wrote several patches in the PR request. You can follow the link. If you don't understand how to do that, just wait.
> 
> You can also give me a link to the WBM model, and I'll try to make it work. I don't know which model you are using, and I would need a link to check your model.



> I wrote several patches in the PR request. You can follow the link. If you don't understand how to do that, just wait.
> 
> You can also give me a link to the WBM model, and I'll try to make it work. I don't know which model you are using, and I would need a link to check your model.

Can you give me the please the link where I find the correction please because I can't find it :( . Thank you

**Comment by akaviaLab** [comment]
The link is here #1197

That error about tuples is because the compartments are weird in harvey4.
If you don't understand what a PR is or how to use it, just wait. The code
will change so harvey4 will work.

**Comment by Ilhem36** [comment]
Thank you for answer .Should I install the last version of cobra to work ?

**Comment by akaviaLab** [comment]
I don't know. It depends.
1) If you have a different format to get this model which works in cobrapy,
just download it and load it from that.
2) If you have access to MATLAB, you can try loading this model into matlab
and exporting it as SBML. Cobrapy should load the SBML correctly.
3) If you don't need the model soon, just wait.
4) If you need the model soon, only have it available in MAT format, don't
have MATLAB, then you can download the new version of the source code. It
might be tricky to run.

**Comment by cdiener** [comment]
Now merged in 0.25.0
